test/compiler/compiler.jl

```diff
@@ -1583,3 +1583,16 @@ end
 
 # Equivalence of Const(T.instance) and T for singleton types
 @test Const(nothing) ⊑ Nothing && Nothing ⊑ Const(nothing)
+
+# issue #27316 - inference shouldn't hang on these
+f27316(::Vector) = nothing
+f27316(::Any) = f27316(Any[][1]), f27316(Any[][1])
+@test Tuple{Nothing,Nothing} <: Base.return_types(f27316, Tuple{Int})[1]
+function g27316()
+    x = nothing
+    while rand() < 0.5
+        x = (x,)
+    end
+    return x
+end
+@test Tuple{Tuple{Nothing}} <: Base.return_types(g27316, Tuple{})[1]
```

Comment:
should make these `==` tests, so we are more likely to be alerted if the test breaks and stops testing what we intended it to test

Reply:
Only if that's actually the right condition. Otherwise an arbitrarily overstrict test with no comment about what to do if it starts failing is pretty bad and will cause future pain and confusion.

Reply:
The test's intention is to verify that inference finishes in a finite amount of time. In this case, it now returns `Any`, which I'm not sure we want to test for explicitly. The lower bounds I'm using instead are chosen somewhat arbitrarily.

Reply:
We can also write this test as `S <: T == Any # we may be able to improve this bound in the future`